src/Client.php

```diff
@@ -32,139 +28,89 @@ class Client implements ClientInterface
 
     /**
-     * @var Event The last event that was captured
+     * @var ReprSerializer The representation serializer
      */
-    private $lastEvent;
+    private $representationSerializer;
 
     /**
      * Constructor.
      *
-     * @param Configuration      $config    The client configuration
-     * @param TransportInterface $transport The transport
+     * @param Options                $options      The client configuration
+     * @param TransportInterface     $transport    The transport
+     * @param IntegrationInterface[] $integrations The integrations used by the client
      */
-    public function __construct(Configuration $config, TransportInterface $transport)
+    public function __construct(Options $options, TransportInterface $transport, array $integrations = [])
     {
-        $this->config = $config;
+        $this->options = $options;
         $this->transport = $transport;
-        $this->tagsContext = new TagsContext();
-        $this->userContext = new UserContext();
-        $this->extraContext = new Context();
-        $this->runtimeContext = new RuntimeContext();
-        $this->serverOsContext = new ServerOsContext();
-        $this->transactionStack = new TransactionStack();
-        $this->serializer = new Serializer($this->config->getMbDetectOrder());
-        $this->representationSerializer = new ReprSerializer($this->config->getMbDetectOrder());
-        $this->middlewareStack = new MiddlewareStack(function (Event $event) {
-            return $event;
-        });
-
-        $request = ServerRequestFactory::fromGlobals();
-        $serverParams = $request->getServerParams();
-
-        if (isset($serverParams['PATH_INFO'])) {
-            $this->transactionStack->push($serverParams['PATH_INFO']);
+        $this->integrations = Handler::setupIntegrations($integrations);
+        $this->serializer = new Serializer($this->options->getMbDetectOrder());
+        $this->representationSerializer = new ReprSerializer($this->options->getMbDetectOrder());
+        if ($this->options->getSerializeAllObjects()) {
+            $this->serializer->setAllObjectSerialize($this->options->getSerializeAllObjects());
+            $this->representationSerializer->setAllObjectSerialize($this->options->getSerializeAllObjects());
         }
+    }
 
-        if ($this->config->getSerializeAllObjects()) {
-            $this->setAllObjectSerialize(true);
-        }
+    /**
+     * {@inheritdoc}
+     */
+    public function getOptions(): Options
+    {
+        return $this->options;
     }
 
     /**
      * {@inheritdoc}
      */
-    public function addBreadcrumb(Breadcrumb $breadcrumb, ?Scope $scope = null)
+    public function addBreadcrumb(Breadcrumb $breadcrumb, ?Scope $scope = null): void
     {
-        $beforeBreadcrumbCallback = $this->config->getBeforeBreadcrumbCallback();
-        $maxBreadcrumbs = $this->config->getMaxBreadcrumbs();
+        $beforeBreadcrumbCallback = $this->options->getBeforeBreadcrumbCallback();
```

Comment:
We're still not handling all the possible types of callback. I've opened #692 to track it after this PR.